src/lib/LocalMedia.ts

```diff
@@ -72,31 +82,76 @@ export default class LocalMedia extends TypedLocalMediaEventTarget {
     }
 
     getCameraDeviceId() {
-        return this.stream.getVideoTracks()[0]?.getSettings().deviceId;
+        return this._currentCameraDeviceId;
     }
 
     getMicrophoneDeviceId() {
-        return this.stream.getAudioTracks()[0]?.getSettings().deviceId;
+        return this._currentMicrophoneDeviceId;
     }
 
     isCameraEnabled() {
-        return !!this.stream.getVideoTracks()[0]?.enabled;
+        return this._cameraEnabled;
     }
 
     isMicrophoneEnabled() {
-        return !!this.stream.getAudioTracks()[0]?.enabled;
+        return this._microphoneEnabled;
     }
 
-    toggleCameraEnabled(enabled?: boolean) {
-        const videoTrack = this.stream.getVideoTracks()[0];
-        if (!videoTrack) {
+    async toggleCameraEnabled(enabled?: boolean) {
+        if (!this._constraints || this._isTogglingCameraEnabled) {
             return;
         }
 
-        const newValue = enabled ?? !videoTrack.enabled;
-        videoTrack.enabled = newValue;
+        let track = this.stream.getVideoTracks()[0];
+        const SHOULD_STOP_TRACK = true; // Stop track in all browsers
 
-        this.dispatchEvent(new CustomEvent("camera_enabled", { detail: { enabled: newValue } }));
+        // Update internal state and dispatch event early
+        this._cameraEnabled = enabled ?? !track?.enabled;
+        this.dispatchEvent(new CustomEvent("camera_enabled", { detail: { enabled: this._cameraEnabled } }));
+
+        this._isTogglingCameraEnabled = true;
+
+        try {
+            if (this._cameraEnabled) {
+                if (track) {
+                    // We have existing video track, just enable it
+                    track.enabled = true;
+                } else {
+                    // We dont have video track, get new one
+                    const newStream = await navigator.mediaDevices.getUserMedia({
+                        video: {
+                            ...(this._currentCameraDeviceId
+                                ? { deviceId: { exact: this._currentCameraDeviceId } }
+                                : null),
+                        },
+                    });
+                    track = newStream.getVideoTracks()[0];
+                    if (track) {
+                        this.stream.addTrack(track);
+                    }
+                }
+            } else {
+                if (!track) {
+                    return;
+                }
+
+                track.enabled = false;
+
+                if (SHOULD_STOP_TRACK) {
```

Comment:
Why do we need this `SHOULD_STOP_TRACK=true` constant and therefore this always `true` condition?

Reply:
Im reworking this to be dependent on whether we have passed an existing stream or not. So if we are passed a stream, we never stop the track, only disable it.

Reply:
Fixed